vec-lib/main.rkt

```diff
@@ -159,42 +151,47 @@
 (define vec-a vec-w)
 
 
-(define (vec-op op)
-  (define (outter lhs rhs . others)
-    (let ([inner
-           (λ (lhs rhs)
-             (cond [(number? lhs)
-                    (for/list ([num (in-list rhs)])
-                      (op lhs num))]
-                   [(number? rhs)
-                    (for/list ([num (in-list lhs)])
-                      (op num rhs))]
-                   [else
-                    (for/list ([i (in-range (min (length lhs) (length rhs)))])
-                      (let ([a (list-ref lhs i)]
-                            [b (list-ref rhs i)])
-                        (op a b)))]))])
-      (let ([ret (inner lhs rhs)])
-        (if (null? others)
-            ret
-            (apply outter (cons ret others))))))
-  outter)
+(define-syntax-rule (define-vec-op name op-expr)
+  ;; macro layer to make `object-name` work w/o runtime overhead
+  (define name
+    (let ([op op-expr])
+      (define (name lhs rhs . others)
+        (perform-vec-op op lhs rhs others))
+      name)))
+(define (perform-vec-op op lhs rhs others)
+  ;; helper function to reduce size of generated code
+  (define ret
+    (cond
+      [(number? lhs)
+       (for/list ([num (in-list rhs)])
+         (op lhs num))]
+      [(number? rhs)
+       (for/list ([num (in-list lhs)])
+         (op num rhs))]
+      [else
+       ;; will stop at the end of the shorter list (unlike `map`)
+       (for/list ([a (in-list lhs)]
+                  [b (in-list rhs)])
+         (op a b))]))
+  (if (pair? others)
+      (perform-vec-op op ret (car others) (cdr others))
+      ret))
```

Comment:
It usually isn't worth thinking about this, but checking `pair?` rather than `null?` ensures that the compiler will be able to optimize `car` and `cdr` to `unsafe-car` and `unsafe-cdr`.